podpac/core/node.py

```diff
@@ -577,8 +577,8 @@ def has_cache(self, key, coordinates=None):
         if self.cache_ctrl is None:
             return False
 
-        with thread_manager.cache_lock:
-            return self.cache_ctrl.has(self, key, coordinates=coordinates)
+        # with thread_manager.cache_lock:
```

Comment:
This keeps biting me with cached properties. It's pretty hard not to hit this lock in a single process. 

I think removing it is also pretty safe, since if multiple threads are reading there's not issue. The only foreseeable problem is if another Node if writing to the cache one one Thread, and a second Node reads it incompletely (reading faster than writing). But this is remote enough that I'm not going to worry about that case right now. 

The right thing to do is to apply the lock in the cache module.

Reply:
Okay. That's too bad. #371 was supposed to help with this, so I'm curious what problems you ran into.

Most of the times I had trouble here, it was because a node attr had a default method that tried to access the cache, and without the lock you get an infinite recursion instead, so removing it didn't help.

> I think removing it is also pretty safe, since if multiple threads are reading there's not issue.

That's true. We should use the correct synchronization tools to allow multiple simultaneous readers.

> The right thing to do is to apply the lock in the cache module.

Yeah, I guess this is also true. It would be more clear what needs to wait for what at that level, too.

Reply:
Summary: I think we should 
 * remove the lock here in the node (as you have done here)
 * reinstate the `trait_is_default` check (see below)
 * make a new issue for adding appropriate synchronization into the cache module (including #368, simultaneous readers, etc)

---

We discussed this over in #368, and I was reminded that one of the things that I did during #371 was to omit node attrs that used the default value from the definition. In addition to looking cleaner to a user, this had the *primary* benefit of never calling `tl.default` methods when caching, which prevented most of the deadlocks that I was encountering.

(Note that it is sufficient to check `trait_is_defined` in order to avoid calling the `tl.default` methods when creating the definition, but that breaks the stability of the node definition and therefore also the node hash.)

The catch is that towards the end of #371, I reverted the `trait_is_default` check. Traitlets handles `List` and `Dict` trait defaults poorly, so it required a bit of hack in a few places. I forgot the primary motivation (deadlocking), and "cleaner node definitions" as a user-exeperience-optization just didn't justify the hacks.


https://github.com/creare-com/podpac/commit/35fe059e16d76dd8ffa12b9b5b297bc0f07076f7 is the commit that removed `trait_is_default` (but a lot of other things), and you can see an example of the "hack" to get traitlets container defaults to work as expected in coord_select.py. I think it would take about an hour to reinstate.

Does that make sense, and so do you agree I should add the `trait_is_default` check back in?

Reply:
Yeah, the compositor `outputs` default really is always the problem. I'm surprised it would hit the thread lock, but doesn't go into an infinite loop without the lock. I'll just have to try it out when I get a chance.

Good point about the defaults changing between versions.

Reply:
Okay. The cosmos stations node currently with or without the thread lock because you have set `outputs=None` for the compositor.

Without setting `outputs=None`, you get either deadlock (with the thread lock) or infinite recursion. I think this is a key point that needs to be fixed, because it is not clear, and it makes the compositor hard to use. I think that the outputs default needs to be None for the Compositor class with an explicit mixin or flag to enable automatic multiple outputs detection. Let me throw you a quick proposal for that.

Reply:
Yikes, I missed the infinite recursion... is there a sane way we can break out of it? I mean, if we managed to set up an infinite recursion, i suspect others will end up doing the same thing. Is it tied to caching? Could we detect a loop in the cache and then just recalculate when a loop is detected?

Reply:
@mpu-creare Please let me know about #387 or #388. I think removing the thread lock here is fine, in either case.